test/core/TypeDBCoreRunner.java

```diff
@@ -0,0 +1,147 @@
+import static com.vaticle.typedb.common.test.RunnerUtil.findUnusedPorts;
+import static com.vaticle.typedb.common.test.RunnerUtil.typeDBCommand;
+import static com.vaticle.typedb.common.test.RunnerUtil.unarchive;
+import static com.vaticle.typedb.common.test.RunnerUtil.waitUntilPortUsed;
+
+public class TypeDBCoreRunner implements TypeDBRunner {
+
+    private final Path distribution;
+    private final Path dataDir;
+    private final Path logsDir;
+    private final int port;
+    private StartedProcess process;
+    private final ProcessExecutor executor;
+
+    public TypeDBCoreRunner() throws InterruptedException, TimeoutException, IOException {
+        port = findUnusedPorts(1).get(0);
+        System.out.println(address() + ": Constructing " + name() + " runner");
+        System.out.println(address() + ": Extracting distribution archive...");
+        distribution = unarchive();
+        System.out.println(address() + ": Distribution archive extracted.");
+        dataDir = distribution.resolve("server").resolve("data");
+        logsDir = distribution.resolve("server").resolve("logs");
+        executor = createProcessExecutor(distribution);
+        System.out.println(address() + ": Runner constructed");
+    }
+
+    private String name() {
+        return "TypeDB Core";
+    }
+
+    @Override
+    public String address() {
+        return host() + ":" + port();
+    }
+
+    public String host() {
+        return "127.0.0.1";
+    }
+
+    public int port() {
+        return port;
+    }
+
+    @Override
+    public void start() {
+        try {
+            System.out.println(address() + ": " +  name() + "is starting... ");
+            System.out.println(address() + ": Distribution is located at " + distribution.toAbsolutePath());
+            System.out.println(address() + ": Data directory is located at " + dataDir.toAbsolutePath());
+            System.out.println(address() + ": Bootup command = " + command());
+            process = executor.command(command()).start();
+            boolean started = waitUntilPortUsed(host(), port())
+                    .await(SERVER_STARTUP_TIMEOUT_MILLIS, TimeUnit.MILLISECONDS);
+            if (!started) {
+                String message = address() + ": Unable to start. ";
+                if (process.getFuture().isDone()) {
+                    ProcessResult processResult = process.getFuture().get();
+                    message += address() + ": Process exited with code '" + processResult.getExitValue() + "'. ";
+                    if (processResult.hasOutput()) {
+                        message += "Output: " + processResult.outputUTF8();
+                    }
+                }
+                throw new RuntimeException(message);
+            } else {
+                System.out.println(address() + ": Started");
+            }
+        } catch (Throwable e) {
+            printLogs();
+            throw new RuntimeException(e);
+        }
```

Comment:
is this all common code with the `TypeDBClusterServerRunner`?

Reply:
Yes they indeed are! Do you have any suggestions on how to not duplicate the code?

Reply:
Maybe the `TypeDBRunner` can/should be `abstract` and abstract away sone of this common stuff? Otherwise not super sure

Reply:
> Maybe the TypeDBRunner can/should be abstract and abstract away sone of this common stuff

It shouldn't be `TypeDBRunner` as the code is only common between `TypeDBCoreRunner` and `TypeDBClusterServerRunner`, but not `TypeDBClusterRunner`.

Reply:
I'll try to refactor it into `RunnerUtil` making sure that the domain-specific stuff are extracted away.

Reply:
Addressed in 9d5f4845d09b59f22c981a106ade05096d9f14c1.